### ISSUE TITLE
remote_state: special notation in attribute (output) names (dots, brackets)

### ISSUE THREAD

**Issue by DeviaVir** [hashicorp/terraform#17197]
### Terraform Version
```
Terraform v0.11.2
```

### Terraform Configuration Files
https://gist.github.com/DeviaVir/4360d4189a10882a13bd95381e2b79b7 and also included below:

```json
[...]
  "output": {
    "Bucket[networking-test-bucket].self_link": {
      "value": "${google_storage_bucket.networking-test-bucket.self_link}"
    },
    "Bucket[networking-test-bucket].url": {
      "value": "${google_storage_bucket.networking-test-bucket.url}"
    }
  },
[...]
```

```json
{
  "data": {
    "terraform_remote_state": {
      "networking-int": {
        "backend": "gcs",
        "config": {
          "bucket": "networking-int",
          "path": "terraform.tfstate",
          "region": "us-east1",
          "project": "networking-int"
        }
      }
    }
  },
  "output": {
    "Bucket[networking-test-bucket].self_link": {
      "value": "${data.terraform_remote_state.networking-int.Bucket[networking-test-bucket].self_link}", // nope
      "value": "${data.terraform_remote_state.networking-int[Bucket[networking-test-bucket].self_link]}", // nope
      "value": "${data.terraform_remote_state.networking-int[\"Bucket[networking-test-bucket].self_link\"]}", // also nope
      "value": "${data.terraform_remote_state.networking-int.[\"Bucket[networking-test-bucket].self_link\"]}", // nope
      "value": "${data.terraform_remote_state.networking-int.\"Bucket[networking-test-bucket].self_link\"}", // nope
      "value": "${element(data.terraform_remote_state.networking-int.*, index(data.terraform_remote_state.networking-int.*, \"Bucket[networking-test-bucket].self_link\"))}", //nope
      "value": "${lookup(data.terraform_remote_state.networking-int.Bucket, \"Bucket[networking-test-bucket].self_link\")}" //nope
    }
  }
}
```

### Debug & Crash Output
Included below the various debug outputs for all of the values I tried:
```
* output.Bucket[networking-test-bucket].self_link: Resource 'data.terraform_remote_state.networking-int' does not have attribute 'Bucket' for variable 'data.terraform_remote_state.networking-int.Bucket'

[...]

Error: Error loading ~/main.tf.json: Error reading config for output Bucket[networking-test-bucket].self_link: data.terraform_remote_state.networking-int: data variables must be four parts: data.TYPE.NAME.ATTR in:

${lookup(data.terraform_remote_state.networking-int, "Bucket[networking-test-bucket].self_link")}

[...]

* output.Bucket[networking-test-bucket].self_link: Resource 'data.terraform_remote_state.networking-int' does not have attribute '*' for variable 'data.terraform_remote_state.networking-int.*'

[...]

Error: Error loading ~/main.tf.json: Error reading config for output Bucket[networking-test-bucket].self_link: parse error at 1:53: expected "}" but found opening quote

[...]

Error: Error loading ~/main.tf.json: Error reading config for output Bucket[networking-test-bucket].self_link: parse error at 1:90: expected "}" but found "."

[...]

* output.Bucket[networking-test-bucket].self_link: Resource 'data.terraform_remote_state.networking-int' does not have attribute '' for variable 'data.terraform_remote_state.networking-int.'

```

### Expected Behavior
I expected there to be a way to reference all of the available outputs, or to use one of the syntaxes to pass. Much in the same fashion that `terraform output -json Bucket[networking-test-bucket].self_link` actually does work.

### Actual Behavior
There seems to be no way to reference data attributes with a special notation.

edit: if a method to use special notation exists, then this could also solve https://github.com/hashicorp/terraform/issues/16843

**Comment by DeviaVir** [comment]
@apparentlymart that sounds great to me, we can't wait!

**Comment by apparentlymart** [comment]
Hi again, @DeviaVir!

I'm pleased to report that the new capabilities I mentioned earlier are now merged into the master branch ready to be included in the forthcoming v0.12.0 release. I've verified a similar configuration with the v0.12.0-alpha1 prerelease build and seen it work, and also double-checked that the new version is indeed reporting an error when an output name includes non-identifier characters.

Since this is now in master, I'm going to close this out. Thanks for reporting this, and for your patience while we did the groundwork to make this possible.
